include/rcutils/filesystem.h

```diff
@@ -185,19 +192,46 @@ bool
 rcutils_mkdir(const char * abs_path);
 
 /// Calculate the size of the specified directory.
-/*
+/**
  * Calculates the size of a directory by summarizing the file size of all files.
  * \note This operation is not recursive.
  * \param[in] directory_path The directory path to calculate the size of.
  * \param[in] allocator Allocator being used for internal file path composition.
- * \return The size of the directory in bytes.
+ * \return The size of the directory in bytes on success.
+ * \return -1 on failure.
  */
 RCUTILS_PUBLIC
-size_t
+ssize_t
 rcutils_calculate_directory_size(const char * directory_path, rcutils_allocator_t allocator);
 
+/// Calculate the size of the specified directory with recursion.
+/**
+ * Calculates the size of a directory and subdirectory by summarizing the file size of all files.
+ * If necessary, you can specify the maximum directory depth to recurse into.
+ * Depth definition as below.
+ * \code
+ * directory_path  <= depth 1
+ *    |- subdirectory <= depth 2
+ *            |- subdirectory <= depth 3
+ *                    ...
+ * \endcode
+ *
+ * \note This operation doesn't follow the symbolic link that points to directory.
+ * \param[in] directory_path The directory path to calculate the size of.
+ * \param[in] max_depth The maximum depth of subdirectory. 0 means no limitation.
+ * \param[in] allocator Allocator being used for internal file path composition.
+ * \return The size of the directory in bytes on success.
+ * \return -1 on failure.
+ */
+RCUTILS_PUBLIC
+ssize_t
+rcutils_calculate_directory_size_with_recursion(
+  const char * directory_path,
+  const size_t max_depth,
+  rcutils_allocator_t allocator);
```

Comment:
I'm very sorry about this, but I realize I made a mistake suggesting an `ssize_t` return value here.

On 64-bit systems (amd64, Aarch64), this will work fine.  We won't be seeing file sizes greater 2^63 - 1.

But on 32-bit systems (x86, arm32), the maximum size we'll be able to return is 2^31 - 1 .  That's just over 2 GB, and it seems entirely plausible that we'll see bags (and other things) larger than that.  While 32-bit systems aren't Tier 1 support, they are something that we try to keep working.  So my suggestion is that we change this signature to:

```
rcutils_ret_t
rcutils_calculate_directory_size_with_recursion(
  const char * directory_path,
  const size_t max_depth,
  uint64_t * size,
  rcutils_allocator_t allocator);
```

What do you think?

Reply:
Yeah. This is a problem.   
I will update codes.